rmf_task_ros2/src/rmf_task_ros2/dispatcher/Dispatcher.cpp

```diff
@@ -0,0 +1,323 @@
+    bid_notice.time_window = rmf_traffic_ros2::convert(
+      rmf_traffic::time::from_seconds(bidding_time_window));
+    auctioneer->start_bidding(bid_notice);
+
+    return submitted_task.task_id;
+  }
+
+  bool cancel_task(const TaskID& task_id)
+  {
+    // check if key exists
+    if (!(*active_dispatch_tasks).count(task_id))
+      return false;
+
+    TaskProfile profile;
+    {
+      // make sure status will expired, todo: cleaner way
+      auto cancel_task_status = (*active_dispatch_tasks)[task_id];
+      profile = cancel_task_status->task_profile;
+
+      // Cancel bidding. This will remove the bidding process
+      if (cancel_task_status->state == TaskStatus::State::Pending)
+      {
+        cancel_task_status->state = TaskStatus::State::Canceled;
+        terminate_task(cancel_task_status);
+
+        if (on_change_fn)
+          on_change_fn(cancel_task_status);
+
+        return true;
+      }
+    }
+
+    // Cancel action task, this will only send a cancel to FA. up to
+    // the FA whether to cancel the task. On change is implemented
+    // internally in action client
+    return action_client->cancel_task(profile);
+  }
+
+  rmf_utils::optional<TaskStatus::State> get_task_state(
+    const TaskID& task_id)
+  {
+    // check if key doesnt exist
+    if ((*active_dispatch_tasks).count(task_id))
+      return (*active_dispatch_tasks)[task_id]->state;
+
+    if ((*terminal_dispatch_tasks).count(task_id))
+      return (*terminal_dispatch_tasks)[task_id]->state;
+
+    return rmf_utils::nullopt;
+  }
+
+  void receive_bidding_winner_cb(
+    const TaskID& task_id,
+    const rmf_utils::optional<bidding::Submission> winner)
+  {
+    if (!(*active_dispatch_tasks).count(task_id))
+      return;
+
+    std::cout << "[Dispatch::BiddingResultCb] | task: " << task_id;
+    auto pending_task_status = (*active_dispatch_tasks)[task_id];
+
+    if (!winner)
+    {
+      std::cerr << " | No winner found :( " <<  std::endl;
+      pending_task_status->state = TaskStatus::State::Failed;
+      terminate_task(pending_task_status);
+
+      if (on_change_fn)
+        on_change_fn(pending_task_status);
+
+      return;
+    }
+
+    // now we know which fleet will execute the task
+    std::cout << " | Found a winner! " << winner->fleet_name << std::endl;
+    pending_task_status->fleet_name = winner->fleet_name;
+
+    // Remove previous self-generated charging task from "active_dispatch_tasks"
+    // this is to prevenet duplicated charging task (as certain queued charging
+    // tasks are not terminated when task is reassigned).
+    // TODO: a better way to impl this
+    for (auto it = active_dispatch_tasks->begin();
+      it != active_dispatch_tasks->end(); )
+    {
+      auto task_type = it->second->task_profile.task_type.type;
+      bool is_charging_task = (task_type == TaskType::TYPE_CHARGE_BATTERY);
+      bool is_fleet_name = (winner->fleet_name == it->second->fleet_name);
+
+      if (is_charging_task && is_fleet_name)
```

Comment:
I'm very very confused about how automatic charging tasks can end up in the `active_dispatch_tasks` map. As far as I can this map would only contain user-requested tasks.

Reply:
I admit this is surely confusing as this is a 'custom fix' for adding the battery charging tasks to the `active_dispatch_tasks`. This is to show it on the GUI and inform the user of a Charging Task.

Having this implementation is due to the fact that the current fleet adapter will self-initiate a `ChargeBattery` task without going through the dispatcher. The `task_id` of the charging task is also self-generated (during reassignment) within the fleet adapter. Definitely, we will need a better way to implement this.

Reply:
Oh, I see. So the fleets will intentionally report when they plan on recharging the battery? That makes sense.

We should definitely work towards a harder distinction between requested tasks and automatic tasks in a future iteration. I've opened an issue here: https://github.com/osrf/rmf_core/issues/241